### ISSUE TITLE
Support for JSDoc interface definition

### ISSUE THREAD

**Issue by stof** [microsoft/TypeScript#33207]
## Search Terms

jsdoc, interface
I found https://github.com/microsoft/TypeScript/issues/16142 which was closed by the reporter, and never reopened even though some other people suggested it.

## Suggestion

JSDoc has a way to declare interfaces, through `@interface`, `@function` (and then `@implements` on the class). It would be great if tsc could support them (defining a TS interface based on that info).

## Use Cases

JSDoc interfaces are useful to define interfaces implemented by multiple classes, that are then used based on this interface. This is part of JSDoc, but currently unsupported by tsc.

## Examples

```js
/**
 * @interface Metric
 */
/**
 * @function
 * @name Metric#getName
 * @returns {string}
 */
/**
 * @function
 * @name Metric#compute
 * @param context
 * @param [extra]
 * @returns {Promise<object>}
 */

/**
 * @constructor
 * @implements {Metric}
 */
function ActivityScoreMetric () {
    // [REDACTED]
}
ActivityScoreMetric.prototype = {
  compute: function (context, extra) {
      // [REDACTED]
  },

  getName: function () {
    return 'activity_score'
  }
}

// Other implementations go there

// Later code deals with `Array<Metric>` for some variables
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by sandersn** [comment]
This is a sub-issue of https://github.com/microsoft/TypeScript/issues/30624. Right now we have decided not to make all of Typescript's features available in JSDoc. Writing 40 lines of metadata in a comment is not much different from putting those same lines in a d.ts file, and we believe the d.ts syntax is much more readable to humans.

**Comment by stof** [comment]
@sandersn I tend to agree that the syntax of TS is more readable. But here, I'm not asking to support TS syntax inside JSDoc (which is what I understand from #30624) but to support existing JSDoc features (so that code already using them in JSDoc can be understood by the TS compiler). This is an existing JSDoc syntax for interface (and yes, again, I totally agree that it is lot less readable that the TS syntax).

**Comment by sandersn** [comment]
I looked for uses of `@interface` and `@implements` by grepping around the dependencies of our user tests. It's a good cross-section of JS code. It doesn't look like anybody uses `@interface` the way jsdoc suggests, although I found a few that use it the way that Closure does. 

Until we have evidence that people use `@interface` the way jsdoc suggests, there's no reason for us to support it.

## react-dom ##

Uses `@implements` for `PooledClass`, which is a declared as a `var` with no jsdoc.

Uses `@interface` as a jsdoc on object literals. It's always a reference to what appears to be a superclass, so I think they're using it to mean `@extends`.

## superagent ##

Uses `@implements ReadableStream`, but doesn't define that interface anywhere.

## ejs ##

Same as superagent, except with `Cache, EscapeCallback, RethrowCallback`

## nise and sinon ##

Uses `@interface` on a class with no implementation:

```ts
  /** @interface */
  function Encoder() {}
  Encoder.prototype = {
    handler: function(stream, code_point) {}
  };
```

I'm pretty sure this is how closure defines interfaces. chrome-devtools-frontend, firebase and protobufjs also work this way. These are all google projects, so it's not surprising.

**Comment by stof** [comment]
> Uses `@implements ReadableStream`, but doesn't define that interface anywhere.

isn't `ReadableStream` defined in the nodejs types ?

**Comment by TimvdLippe** [comment]
We use `@interface` extensively in Chrome DevTools and the lack of support with `checkJs` means we can not use TypeScript to check our JavaScript. Our usage: https://cs.chromium.org/search/?q=f:devtools-frontend+@interface&sq=package:chromium&type=cs

I am not sure if we can use the `.d.ts` trick, as TS itself complains about how classes annotated with `@interface` are defined (it complains about missing return statements for example)

**Comment by sandersn** [comment]
@TimvdLippe Could you open a separate issue for supporting closure's `@interface`? It has (1) a different audience (2) different requirements (3) different implementation. Having a separate issue is useful for (1), since we use upvotes as data for prioritising features. But d.ts-based workarounds are different too; I'd like to know what you've tried so far and what Typescript errors you're getting.

**Comment by benmccann** [comment]
We use `@interface` in Chart.js. For example to document our plugin interface: https://github.com/chartjs/Chart.js/blob/771fe520957199b32a26205c14f4816002842bcb/src/core/core.plugins.js#L173

**Comment by jimmywarting** [comment]
I simply just don't use it cuz you don't support it, if it would then i would start using it more... please add support for this.

**Comment by spillz** [comment]
Just to clarify, is something like this use of `@interface` deliberately not a supported use of JSDoc declarations in `tsc`? 

I've been trying to get VS Code to correctly recognize the `Person` interface in function args like below and I assume it's either a tsc or language server that disallows this usage.

```js
//@ts-check

/**
 * @interface Person
 * @property {string} name - The person's name.
 * @property {number} age - The person's age.
 */

/**
 * Greets a person.
 * 
 * @param {Person} person - The person to greet. //VS Code flags this as an error "Cannot find name 'Person'"
 * @returns {string} A greeting message.
 */
function greet(person) {
  return `Hello, ${person.name}! You are ${person.age} years old.`;
}

//Should be a valid call even though occupation is not part of the interface
greet({name:'Joe', age:25, occupation:'miner'});
//Should be an invalid call because age value is wrong type
greet({name:'Joe', age:'X', occupation:'miner'});
```
